File: code-monkey/src/tools/read_file_tool.py
import logging
import traceback
from pydantic import BaseModel, Field
from typing import Type, Optional
from langchain_core.tools import BaseTool
from langchain_core.callbacks import (
    AsyncCallbackManagerForToolRun,
)
from .utils import make_file_path
from instrumentation import instrument

class ReadFileToolInput(BaseModel):
    fname: str = Field(description="The filename to read")

class ReadFileTool(BaseTool):
    """Tool that returns the contents of a file given the filename"""
    name: str = "read_file"
    description: str = "Read the contents of the file of given name"
    args_schema: Type[BaseModel] = ReadFileToolInput

    @instrument("Tool._run", ["fname"], attributes={ "tool": "ReadFileTool" })
    def _run(self, fname: str, run_manager: Optional[AsyncCallbackManagerForToolRun] = None) -> str:
        file_path = make_file_path(fname)
        try:
            with open(file_path, "r") as file:
                content = file.read()
            return content
        except Exception:
           logging.error("Failed to open file for reading: %s", file_path)
           traceback.print_exc()
           # Re-raise the exception
           raise

File: code-monkey-ts2/tools/read_file_tool.ts
import { z } from "zod";
import { readFileContent } from "./utils";
import { instrument, currentSpan } from "../instrumentation";
import { IOTool } from './io_tool';
import { CodeContext } from "../code_context";

const schema = z.object({
  fname: z.string().describe("Name of the file to edit."),
});

export class ReadFileTool extends IOTool {
  name = "read_file";
  description =
    "Read the contents of the file of given name.";
  schema = schema;

  constructor(codeContext: CodeContext) {
    super(codeContext);
  }

  @instrument("Tool._call", { tool: "ReadFileTool" })
  async _call({ fname }: z.infer<typeof schema>): Promise<string> {
    currentSpan().setAttributes({
      fname,
    });
    const filePath = this.codeContext.resolveFile(fname);
    try {
      const content = await readFileContent(filePath);
      return `${content}`;
    } catch (error) {
      // console.error(`Failed to open file for reading: ${filePath}`);
      // console.error(error);
      return `Failed to read file: ${error?.stack || error}`;
    }
  }
}